[WIP] `async` and `.await` chapter (#3909)
rust-lang/book: src/ch17-01-futures-and-syntax.md

async programming. (If you are familiar with other languages’ approach to async,
you may notice some significant differences, though.) In Rust, blocks and
functions can be marked `async`, and you can wait on the result of an `async`
function or block to resolve using the `await` keyword.

Let’s write our first async function, and call it:

<Listing number="17-1" file-name="src/main.rs" caption="Defining a very simple async function">

```rust
{{#rustdoc_include ../listings/ch17-async-await/listing-17-01/src/main.rs:all}}
```

</Listing>

If we compile and run this… nothing happens, and we get a compiler warning:

```console
{{#include ../listings/ch17-async-await/listing-17-01/output.txt}}
```

The warning tells us that just calling `hello` was not enough: we also need to
`.await` or poll the future it returns. This raises some important questions:

- Given there is no return type on the function, how is it returning a future?
- What exactly is a future?
- Why do we need to `.await` or poll futures to make them do something?
- How do `.await` and polling relate to each other?

We will work through each of these in turn. We can answer the first question by

> learning what the syntax means, so let’s start there.

Reviewer: Concur: I am rewriting this entire section by replacing the opening example!

### Async functions

In Rust, writing `async fn` is equivalent to writing a function which returns a
*future* of the return type. That is, when the compiler sees a function like
`async fn hello` in Listing 17-1, it is basically equivalent to a function
defined like this instead:

```rust
fn hello(name: &str) -> impl Future<Output = ()> {
    async {
        let greeting = format!("Hello, {name}!");
        println!("{greeting}");
    }
}

> ```

Reviewer 2: These two sentences should be consistent:

> ...is equivalent to ... 
>
> ... is basically equivalent to ...

The difference can seed doubt in some readers' minds.

Some questions a beginner might ask after the snippet:

- What is a future? I don't think that's been defined yet.
- Where did `Future` come from?
- Is there a difference between "a *future*" (the concept?) and "a `Future`" (the trait?)

I would add a note in this paragraph that says that the next syntax introduces some new elements which will be explained later.

Let’s walk through each part of the transformed version:

* It uses the `impl Trait` syntax we discussed back in the [“Traits as
  Parameters”][impl-trait] section in Chapter 10.
* The returned trait is a `Future`, with an associated type of `Output`. Notice
  that the `Output` type is `()`, which is the same as the the original return

>   type from the `async fn` version of `hello`.

Reviewer: Yeah, I went back and forth on this specifically *multiple* times, and ended up with this approach because it gave me a way to build up the pieces incrementally, but if we come up with some small example that is *actually using* async, that would be great.

> * The whole body of the function is wrapped in an `async` block. Remember that
>   blocks are expressions. This whole block is the expression returned from the
>   function.

Reviewer 2: ```suggestion
* The whole body of the function wraps an `async` block. Remember that
  blocks are expressions. This whole block is the expression returned from the
  function. `async` blocks return a type that implements `Future`, so the whole
  function returns  a type that implements `Future`.
```

Reviewer: 👍🏼 133bb617

* The async block itself has the “unit” value `()`, since it ends with a
  `println!` statement. That value matches the `Output` type in the return type.

An `async` block corresponds to a data type which implements the `Future` trait,
and the result of the async block will be the `Output` of the `Future`. Thus, an
`async fn`’s return type is an anonymous data type the compiler creates for us,
which implements `Future`. The associated `Output` type for the `Future` is the
return type of the original `async fn`. Thus, calling `hello` in Listing 17-1
returned a `Future<Output = ()>`.

Then Rust warned us that we did not do anything with the future. This is because

> futures are *lazy*: they don’t do anything until you ask them to. This should

Reviewer 2: Should we mention how?

```suggestion
futures are *lazy*: they don’t do anything until you ask them to with `await`. This should
```

might be used to if you have used async in languages like JavaScript or C#. Rust
chose this because it makes chains of methods much nicer to work with.) In
Listing 17-2, we add `.await` to the `hello` call in `main`.

<Listing number="17-2" caption="Attempting to fix a compiler warning by awaiting a future" file-name="src/main.rs">

```rust,ignore,does_not_compile
{{#rustdoc_include ../listings/ch17-async-await/listing-17-02/src/main.rs:main}}
```

</Listing>

Oh no! We have gone from a compiler warning to an actual error:

```console
{{#include ../listings/ch17-async-await/listing-17-02/output.txt}}
```

This time, the compiler is informing us we cannot use `.await` in `main`,
because `main` is not an `async` function. Your first thought might be to make
`main` an async function then, as in Listing 17-3.

<Listing number="17-3" caption="Attempting to make `main` an `async fn`" file-name="src/main.rs">

```rust,ignore,does_not_compile
{{#rustdoc_include ../listings/ch17-async-await/listing-17-03/src/main.rs:main}}
```

</Listing>

However, we get another compiler error here:

```console
{{#include ../listings/ch17-async-await/listing-17-03/output.txt}}
```

> The problem is that async code needs a *runtime*: a Rust crate which manages the
> details of executing asynchronous code.

Reviewer 2: I think this needs a little more context. The phrase "The problem" is actually quite hard for learners, because they lack the mental model that you do and are unsure that the problem that they're thinking about is what you're thinking about.

```suggestion
Rust won't allow us too mark `main` as `async`. The underlying problem is that async code needs a *runtime*: a Rust crate which manages the
details of executing asynchronous code. A program's `main` function can initialize a runtime, but it is not a runtime itself.
```

You should also point out that `main` can't be `async`

Reviewer: Concur about the problem here—I do point out that `main` cannot be async, but (a) way too late relative to this paragraph, and (b) there are bigger issues related to the overall structural problems with this section for its audience (which I think you and Carol both hit on!). 🤔

Reviewer: Integrated this exact suggestion directly!

Most languages which support async bundle a runtime with the language. At least

> for now, Rust does not. Instead, there are many different async runtimes

Reviewer 2: > At least for now, Rust does not.

I don't recommend speculating about future Rust.

Reviewer: 👍🏼 133bb617

available, each of which makes different tradeoffs suitable to the use case they
target. For example, a high-throughput web server with dozens of CPU cores and

> terabytes of RAM has very different different needs than a microcontroller with

Reviewer 2: There are only a few sites in the world running web servers with terabytes of RAM per node.

Reviewer 3: Hetzner offers servers with 512GB of RAM from a quick look, it would certainly be beefy for a web server though :)

Reviewer: Ha! Yes, I think my brain subbed in storage sizes. 😂

Reviewer: 👍🏼 133bb617

a single core, one gigabyte of RAM, and no ability to do heap allocations.

> ### The `trpl` Crate
>
> To keep this chapter focused on learning async, rather than juggling parts of
> the ecosystem, we have created the `trpl` crate (`trpl` is short for “The Rust
> Programming Language”). It re-exports all the types, traits, and functions you
> will need, primarily from the [`futures`][futures-crate] and [`tokio`][tokio]
> crates.
>
> - The `futures` crate is an official home for Rust experimentation for async
>   code, and is actually where the `Future` type was originally designed.
>
> - Tokio is the most widely used async runtime in Rust today, especially (but
>   not only!) for web applications. There are other great runtimes out there,
>   and they may be more suitable for your purposes. We use Tokio under the hood

> >   for `trpl` because it good and widely used.

Reviewer 4: "it good and widely used" - typo

Reviewer: it good tho! and widely used! 😂 Thank you, good catch.

Reviewer: 👍🏼 133bb617

> extensive comments explaining what the crate does.

Go ahead and add the `trpl` crate to your `hello-async` project:

```console
$ cargo add trpl
```

Then, in our `main` function, let’s wrap the call to `hello` with the
`trpl::block_on` function, which takes in a `Future` and runs it until it
completes. Since `hello` returns a `Future`, we could simply wrap it directly in
`trpl::block_on`. However, for most of the examples in the chapter, we will be
doing more than just one async function call, so instead we will pass an `async`
block and explicitly await the result of calling `hello`.

<Listing number="17-4" caption="Using the `block_on` helper function to wait on a future in non-async code" file-name="src/main.rs">

```rust
{{#rustdoc_include ../listings/ch17-async-await/listing-17-04/src/main.rs:main}}
```

</Listing>

When we run this, we get the behavior we might have expected initially:

```console
{{#include ../listings/ch17-async-await/listing-17-04/output.txt}}
```

Phew: we finally have some working async code! Now we can turn our attention to
how the `Future` trait works.

### What Are Futures?

> A *future* is a data structure which manages the state of some async operation.

Reviewer 2: "...manages the state of some async operation" is quite a mouthful. can it be expressed in terms that are more beginner-friendly?

> A future is a value that may not ready _yet_. Every future holds its own information about the progress that has been made and what "ready" means. 
>
> In the Rust community, we say that types which implement the `Future` trait are futures.

Reviewer: I like this phrasing a *lot* better than mine; and per Carol’s note about needing to define it right up front (concur!) I will probably steal this and put it there!

Rust provides the `Future` trait as a building block so different async
operations can be implemented with different data, but with a common interface.
Here is the definition of the trait:

```rust
pub trait Future {
    type Output;

    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output>;

> }

Reviewer 2: This trait is scary.

Could you build up to it by creating a series of incomplete traits that motivate each piece? Without more guidance, you'll just end up getting the book slammed shut. 

People using TypeScript don't need to bother with scary traits and we risk sending people back to other communities with the impression that Rust is impenetrable.

There will be plenty of web developers who will skip strait to this chapter and then get stuck on the most complex syntax and semantics in the language.

Reviewer: Yeah, I debated this as well, and I’m entirely open to pulling this chunk out entirely or relocating it. The reason I kept it so far was to give enough of the info for the later discussion of pinning to make sense. I would *love* to not have to talk about `Pin` at *all*, but it is basically impossible to avoid—I tripped over needing to `pin!` things in nearly every small piece of example code I came up with! (I do think the Async WG will have a bunch of that fixed… by the time the 2027 Edition rolls around.)

@timClicks, I *very* much understanding point about people using TS needing to bother with “scary traits” especially, and it’s the biggest reason I debated leaving this out of the chapter entirely. I think (a) that’s *true*, but (b) we also *do* explain at least the basics of `Iterator` in Rust (including in this book) and people in TS usually don’t directly think about `Iterator`, `Iterable`, and `IterableIterator` either (and oh *wow* is that a set of APIs!).  There’s a balance to be struck here, and I definitely don’t claim to have gotten the right one—in fact, I’m sure I leaned into the “explain the machinery!” because that’s how I tend to learn best.

In terms of my actual goals with this material: I want to give people *just enough* that when they hit the real-world consequences of the design of this trait, they can have a sense of what to do next; and I want to give people a clear sense that none of this is magic—as @timClicks put it in a different comment, which I also agree with, it’s ultimately just some traits and nice syntax to make them easier to work with!

---

A thought: I wonder if inverting the order would help—i.e. *motivate it* by way of the pinning discussion, and show this *then*. Then it’s not a “WHOA THERE I JUST WANTED TO WRITE SOME ASYNC CODE, THIS IS TOO MUCH, PEACE OUT” but instead a bit of context to explain why we need to solve a particular problem.

Reviewer: Circling back to this, I concluded that you were basically just correct, @timClicks and @carols10cents. 😂 I pulled the trait definition out entirely and moved it into a new section near the end of the chapter. It may *still* need some of that “incremental building up”, though as a section which is explicitly doing more of a “deeper dive” it may be more okay to just jump in. Either way, the result is that it’s *not* there earlier in the chapter and I think the potential “scary factor” is much lower. 💙

implement `Future` on your own data types. Indeed, many of the functions we will
see throughout this chapter return types with their own implementations of
`Future`.

As we learned earlier, `Future`’s associated type `Output` says what the future
will resolves to. (This is analogous to the `Item` associated type for the
`Iterator` trait.) Beyond that, `Future` also has the `poll` method, which takes
a special `Pin` reference for its `self` parameter and a mutable reference to a
`Context` type, and returns a `Poll<Self::Output>`. We will talk a little more
about `Pin` and `Context` later in the chapter. For now, let’s focus on what the
method returns, the `Poll` type:

```rust
enum Poll<T> {
    Ready(T),
    Pending
}
```

This `Poll` type is a lot like an `Option`: it has one variant which has a value
(`Ready(T)`), and one which does not (`Pending`). It means something quite
different, though! The `Pending` variant indicates that the future still has
work to do, so the caller will need to check again later. The `Ready` variant
indicates that the `Future` has finished its work and the `T` value is
available.

> > Note: With most futures, the caller should not call `poll()` again after the

Reviewer: Ah, good callout. I think *if* we do that back when talking about iterators yes; otherwise I should probably just remove it here (just so the level of detail we get into is comparable between the two).

> future has returned `Ready`. Many futures will panic if polled again after
> becoming ready! Futures which are safe to poll again will say so explicitly in
> their documentation.

Under the hood, when you call `.await`, Rust compiles that to code which calls
`poll`, kind of (although not exactly <!-- TODO: describe `IntoFuture`? -->)
like this:

```rust,ignore
match hello("async").poll() {
    Ready(_) => {
        // We’re done!
    }
    Pending => {
        // But what goes here?
    }
}
```

> What should we do when the `Future` is still `Pending`? We need some way to try

Reviewer: Do Not™ with capital D, capital N, and about seventeen ™s. 😂 I think your instinct here is probably the right one for the audience, though I admit it pains me because I *want* to provide “mental machinery”—I find it really frustrating to just kind of glide over this stuff when learning, but I think I am not representative!

again… and again, and again, until the future is finally ready. In other words,
a loop:

```rust,ignore
let hello_fut = hello("async");
loop {
    match hello_fut.poll() {
        Ready(_) => {
            break;
        }
        Pending => {
            // continue
        }
    }
}
```

If Rust compiled it to exactly that code, though, every `.await` would be
blocking—exactly the opposite of what we were going for! Instead, Rust needs
makes sure that the loop can hand off control to something which can pause work
on this future and work on other futures and check this one again later. That
“something” is an async runtime, and this scheduling and coordination work is
one of the main jobs for a runtime.

Every *await point*—that is, every place where the code explicitly applies the
`.await` keyword—represents a place where control gets handed back to the
runtime. To make that work, Rust needs to keep track of the state involved in
the async block, so that the runtime can kick off some other work and then come
back when it is ready to try advancing this one again. This is an invisible
state machine, as if you wrote something like this:

```rust

> enum MyAsyncStateMachine {

Reviewer 4: I think this example is a little abstract without fleshing it out. There would be value in actually writing down how the state machine would look. That would especially help in later understanding `Pin`.

    FirstAwaitPoint {
        // the state used up to the first await point...
    },
    SecondAwaitPoint {
        // the state used up to the second await point...
    },
    // etc. for each `.await` point...
}
```

Writing that out by hand would be tedious and error-prone, especially when
making changes to code later. Instead, the Rust compiler creates and manages the
state machine data structures for async code automatically.

If you’re wondering: yep, the normal borrowing and ownership rules around data

> structures all apply. Happily, the compiler also handles checking those for us,

Author: _This_ to me is the important part that people who call async code need to understand -- an async block needs to be able to hold onto all the data it needs while it's paused, and know everything it needs will be available when it's resumed, and Rust's ownership/borrowing rules are what ensure that.

and has good error messages. We will work through a few of those later in the
chapter!

Now we can understand why the compiler stopped us from making `main` itself an
async function in Listing 17-3. If `main` were an async function, something else
would need to call `poll()` on whatever `main` returned, but main is the
starting point for the program! Instead, we use the `trpl::block_on` function,
which sets up a runtime and polls the `Future` returned by `hello` until it
returns `Ready`.

Every async program in Rust has at least one place where it sets up a runtime and
executes the futures. Those runtimes also often supply async versions of common
functionality like file or network I/O.

> Note: We have skipped over some interesting implementation details in this

> > discussion, because you should not have to think about them when writing Rust.

Author: No you have not!!! LOL IMO you could be skipping over so much more to get to actually writing some async Rust 😁

Reviewer: I mean, I could have said 4× this much, so it felt to *me* like I skipped a lot. 🤣
